core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala

```diff
@@ -0,0 +1,89 @@
+    new SnapshotReferenceCreationValidator(workspaceContext, snapshot)
+}
+
+// thrown to disallow creation of a snapshot reference on an unsupported platform
+class UnsupportedPlatformException(message: String) extends RawlsException(message)
+// thrown to disallow creation of a snapshot reference across cloud boundaries
+class PlatformBoundaryException(message: String) extends RawlsException(message)
+// thrown to disallow creation of a snapshot reference across a protected data boundary
+class ProtectedDataException(message: String) extends RawlsException(message)
+
+class SnapshotReferenceCreationValidator(val workspaceContext: Workspace, val snapshot: SnapshotModel) {
+
+  // TODO: get this information from a more authoritative source rather than relying on the hardcoded bucket prefix
+  private def isWorkspaceProtected: Boolean = workspaceContext.bucketName.startsWith("fc-secure")
+
+  private def isSnapshotProtected: Boolean = snapshot.getSource.asScala.exists(_.getDataset.isSecureMonitoringEnabled)
+
+  private def snapshotPlatform: SnapshotCloudPlatform = {
+    val snapshotSourceDatasets = snapshot.getSource.asScala
+    val datasetCloudPlatforms = snapshotSourceDatasets.map(_.getDataset.getCloudPlatform).filter(_ != null).toSet
+
+    // assume GCP if datasets don't come from a source explicitly marked with another platform
+    if (datasetCloudPlatforms.isEmpty) {
+      return SnapshotCloudPlatform.GCP
+    }
+
+    // fail if datasets sources somehow span multiple clouds
+    if (datasetCloudPlatforms.size > 1) {
+      throw new IllegalStateException(
+        s"snapshot ${snapshot.getId} unexpectedly has more than one cloudPlatform: $datasetCloudPlatforms"
+      )
+    }
+
+    // fail if dataset sourced from something other than GCP aren't all explicitly marked as such
+    // as this might indicate co-mingled GCP (blank) & non-GCP datasets
+    val onlyCloudPlatform = datasetCloudPlatforms.take(1).last
+    onlyCloudPlatform match {
+      case SnapshotCloudPlatform.GCP => onlyCloudPlatform
+      case _ =>
+        if (snapshotSourceDatasets.forall(_.getDataset.getCloudPlatform == onlyCloudPlatform)) onlyCloudPlatform
+        else
+          throw new IllegalStateException(
+            s"snapshot ${snapshot.getId} unexpectedly has sources from more than one cloudPlatform"
+          )
+    }
+
+  }
+
+  // Ideally this would rely on Terra Policy Service, but until TPS is enabled for GCP
+  // We'll have to use this workaround for identifying protected status.
+  @throws(classOf[ProtectedDataException])
+  def validateProtectedStatus(): Unit =
+    if (!isWorkspaceProtected && isSnapshotProtected) {
+      throw new ProtectedDataException("Unable to add protected snapshot to unprotected workspace.")
+    }
+
+  @throws(classOf[UnsupportedPlatformException])
+  @throws(classOf[PlatformBoundaryException])
+  def validatePlatforms(workspacePlatform: WorkspaceCloudPlatform): Unit = {
```

Comment:
see my other comment - it feels slightly off that the `SnapshotReferenceCreationValidator` constructor accepts a `Workspace`, but this method requires a `WorkspaceCloudPlatform` argument … potentially, you can read the platform right from the Workspace object and not need the argument.